With the access to behavior data about Megaline subscribers who have already switched to the new plans . For this classification task, i need to develop a model that will pick the right plan. 
Develop a model with the highest possible accuracy. In this project, the threshold for accuracy is 0.75. Checking the accuracy using the test dataset.

In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

df = pd.read_csv('')

Importing the models to be used later in the code and reading the file with the dataset.

In [2]:
display(df.head(5))

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

Displaying the content of the data for the first five rows.

In [3]:
df.dtypes

calls       float64
minutes     float64
messages    float64
mb_used     float64
is_ultra      int64
dtype: object

Showing the datatypes of the columns in the dataset. The target column is an integer while the others are float.

In [4]:
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
calls       3214 non-null float64
minutes     3214 non-null float64
messages    3214 non-null float64
mb_used     3214 non-null float64
is_ultra    3214 non-null int64
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


None

All fields are are complete. There are no null values.

In [5]:
df.is_ultra.value_counts()

0    2229
1     985
Name: is_ultra, dtype: int64

The number of 0s and 1s are not equal.

Splitting the data into train, valid and test sets in the ratio 3.1.1.

In [6]:
# Split target and features
target = df['is_ultra']
features = df.drop('is_ultra', axis = 1)

# Split into sets
features_train, features_temp, target_train, target_temp = train_test_split(features, target, test_size=0.2,random_state=42)
features_valid, features_test, target_valid, target_test = train_test_split(features_temp, target_temp, test_size=0.2,random_state=42)

Students Comments: what does 'temp' mean? What ratio does the split method divide the data into in this case? 
    I have noticed that using this style of splitting the difference between the training and the test has greatly increased. Am concerned not to heve an overfit model. Please guide

Student's Comment2: Well understood. Thanks.

Defining the features ie the column variables to be used and targets ie the feature used to make predictions, for the 3 datasets.

In [7]:
for estim in range(1,5,1):
        model = DecisionTreeClassifier(random_state=54321)

        model.fit(features_train,target_train)
        predictions_valid = model.predict(features_valid)
        accuracy = accuracy_score(target_valid, predictions_valid)
        
        model.score(features_train, target_train)
        
        print("n_estimators =", estim, ":", accuracy)

n_estimators = 1 : 0.7587548638132295
n_estimators = 2 : 0.7587548638132295
n_estimators = 3 : 0.7587548638132295
n_estimators = 4 : 0.7587548638132295


Defining the model ie DecisionTree classifier inside a loop that specifies the hyperparameter range to be used. Intiate the training using the fit() method.validating the trained model using the valid dataset the checking the accuracy of the valid dataset that has been predicted with the trained model. Then printing the accuracy score.

In [8]:
model = LogisticRegression(random_state=12345, solver='liblinear') 

model.fit(features_train,target_train)

accuracy = accuracy_score(target_valid, predictions_valid)

print("n_estimators =", estim, ":", accuracy)

n_estimators = 4 : 0.7587548638132295


Defining the model ie Logistic Regression. Intiate the training using the fit() method.Getting the accuracy using the validation data. Then printing the accuracy score.

In [19]:
for estim in range(5,16,5):
        for depth in range(1,15,1):
            for min_samples_leaf in range(1,10):

                    model = RandomForestClassifier(random_state=54321, n_estimators=estim, max_depth=depth,min_samples_leaf=min_samples_leaf,min_samples_split=5)

                    accuracy = accuracy_score(target_valid, predictions_valid)

                    model.fit(features_train,target_train)
                    predictions_train = model.predict(features_train)
                    predictions_valid = model.predict(features_valid)
                    predictions_test = model.predict(features_test)

                    #print("n_estimators =", estim, ":","", accuracy)
                    print('Max Depth ' + str(depth)+': ' + str(accuracy) )
                    print()
                    print('min_samples_leaf ' + str(min_samples_leaf)+': ' + str(accuracy) )

Max Depth 1: 0.811284046692607

min_samples_leaf 1: 0.811284046692607
Max Depth 1: 0.7568093385214008

min_samples_leaf 2: 0.7568093385214008
Max Depth 1: 0.7568093385214008

min_samples_leaf 3: 0.7568093385214008
Max Depth 1: 0.7568093385214008

min_samples_leaf 4: 0.7568093385214008
Max Depth 1: 0.7568093385214008

min_samples_leaf 5: 0.7568093385214008
Max Depth 1: 0.7568093385214008

min_samples_leaf 6: 0.7568093385214008
Max Depth 1: 0.7568093385214008

min_samples_leaf 7: 0.7568093385214008
Max Depth 1: 0.7568093385214008

min_samples_leaf 8: 0.7568093385214008
Max Depth 1: 0.7568093385214008

min_samples_leaf 9: 0.7568093385214008
Max Depth 2: 0.7568093385214008

min_samples_leaf 1: 0.7568093385214008
Max Depth 2: 0.7840466926070039

min_samples_leaf 2: 0.7840466926070039
Max Depth 2: 0.7840466926070039

min_samples_leaf 3: 0.7840466926070039
Max Depth 2: 0.7840466926070039

min_samples_leaf 4: 0.7840466926070039
Max Depth 2: 0.7840466926070039

min_samples_leaf 5: 0.78404669260

Max Depth 9: 0.8448074679113186


Defining the model ie RandomForest classifier inside a loop that specifies the hyperparameter range to be used. Intiate the training using the fit() method.validating the trained model using the valid dataset the checking the accuracy of the valid dataset that has been predicted with the trained model. Then printing the accuracy score.

The hyperparameters that are defined by the n_estimators,depth and min_sample_leaf. Their values are defined in the range of the loop for each hyperparameter. From the results it can be noted that from estimators 10 there is a gradual increase in accuracy upto estimator 40 which gives the peak accuracy. From there the accuracy begins to drop with the increase of the estimators.For all other hyperparameters the curve is the same and the highest value in each has been updated in the final_model oof the test_run.

In [15]:
accuracy = accuracy_score(target_valid, predictions_valid)

model.fit(features_train,target_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=14, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=9, min_samples_split=5,
                       min_weight_fraction_leaf=0.0, n_estimators=15,
                       n_jobs=None, oob_score=False, random_state=54321,
                       verbose=0, warm_start=False)

Selecting the best model.

In [16]:
final_model = RandomForestClassifier(n_estimators=5, random_state=54321, max_depth=6,min_samples_leaf=3,min_samples_split=5)
final_model.fit(features_train, target_train)
predictions_test = model.predict(features_test)

accuracy = accuracy_score(target_train, predictions_train)
test_accuracy = accuracy_score(target_test, predictions_test)

Defining the best model as the final model to be used with the test dataset. Then checking the accuracy of the trained and the test datasetafter the predictions.

In [17]:
print("Accuracy")
print("Training set:", accuracy)
print("Test set:", test_accuracy)

Accuracy
Training set: 0.8448074679113186
Test set: 0.7751937984496124


Printing the accuracy results.

Overall conclusion: The best model for the dataset in use is Random Forest, where the training test has an acuracy of 0.84 and the accuracy of 0.77 for the test set.